### ISSUE TITLE
Partial support and notes

### ISSUE THREAD

**Issue by foolip** [web-platform-dx/web-features#915]
Both BCD and caniuse have a concept of partial support. Examples:
https://developer.mozilla.org/en-US/docs/Web/API/ScreenOrientation/lock#browser_compatibility
https://caniuse.com/webp

We have already accumulated a number of features that could benefit from the same:

https://github.com/web-platform-dx/web-features/blob/477f11f33e2966cbd15575ac733178332ee51e51/feature-group-definitions/avif.yml#L6

https://github.com/web-platform-dx/web-features/blob/477f11f33e2966cbd15575ac733178332ee51e51/feature-group-definitions/js-modules.yml#L12-L14

https://github.com/web-platform-dx/web-features/blob/477f11f33e2966cbd15575ac733178332ee51e51/feature-group-definitions/webp.yml#L6

https://github.com/web-platform-dx/web-features/pull/912 is another case currently in review.

In BCD, `partial_implementation` must come with a note, and we should probably do the same.

**Comment by atopal** [comment]
What's the implication for the Baseline status of those items? Is that something we need to assess on a case by case basis or does it make sense to consider partial implementations generally not baseline?

**Comment by foolip** [comment]
I think partial implementation should mean not Baseline, so we for things we consider partially supported we should currently mark them as not supported and add TODOs. This data wouldn't be for more accurate Baseline data, but to capture nuance accurately if web-features is imported to caniuse.com.

**Comment by foolip** [comment]
Note that there are things that BCD considers `partial_implementation` that I think we should still override and say it's not important enough for the whole feature to be considered partial. A missing `onfoo` event handler property is the most obvious such case, where it makes sense to point out at the most granular level in BCD, but it's at most worthy of a note for the overall feature.

**Comment by ddbeck** [comment]
> I think partial implementation should mean not Baseline,

I just wanted to chime in to say that I think this is correct. I think "partial" ought to mean "`baseline: false` with annotations" not a "half Baseline" or something.

**Comment by mirisuzanne** [comment]
Short term, I think it's more helpful to say 'this sub-feature is fully supported' when browsers are consistent in what part they support. If/when browsers do extend the feature, they will discuss support for the new sub-feature separately.

Long term it might make sense to merge sub-features back together into a complete feature - but it seems to me like 'partial support' is only real useful as a flag when the sub-parts can't be easily distinguished. And there are many cases where this merging happens over time without anything being considered 'partial' along the way. When a new option is added to an old feature, we give that new option a new name - we don't un-baseline the existing functionality.

I would vote to unblock this.

**Comment by bramus** [comment]
I’ve found my way to here through https://github.com/GoogleChrome/webstatus.dev/issues/1141, which is about popover that was Baseline, but then wasn’t, but now is again.

I think that the reason why the feature was initially considered to be Baseline is because the `partial_implementation` field from BCD was ignored – something that is issue right here is also touching upon.

Before https://github.com/mdn/browser-compat-data/pull/25787 – which marks the feature a included in Safari iOS 18.3 – the BCD entry for `popover` looked like this:

```json
            "safari_ios": {
              "version_added": "17",
              "partial_implementation": true,
              "notes": "On iOS and iPadOS, popovers are not dismissed when the user taps outside of the popover area, see [bug 267688](https://webkit.org/b/267688)."
            },
```

Had the `partial_implementation` field been taken into account when computing the Baseline status, the feature would never have been marked as Baseline Newly Available.

Looking at the source of `compute-baseline`, this function needs to be adjusted:

https://github.com/web-platform-dx/web-features/blob/b44c75ca30a98706bcfbc61225b51fde3c5b4db5/packages/compute-baseline/src/browser-compat-data/feature.ts#L124-L130

**Comment by ddbeck** [comment]
Replying to @bramus, specifically on this point:

> Had the `partial_implementation` field been taken into account when computing the Baseline status, the feature would never have been marked as Baseline Newly Available.

We do take `partial_implementation` into account. Right now, the calculation treats all `partial_implementation`s from BCD as if they were wholly unsupported. This fails to expose potentially useful information to developers (e.g., you wouldn't know that you can't trust feature detection for popover). But in the absence of notes (or more structured data from BCD), it's the appropriately conservative course to take—we can automatically judge the severity of a partial.

That said, the source comment is very badly written (by me) and can be read to mean the exact opposite of what it does. I've sent https://github.com/web-platform-dx/web-features/pull/2620 to correct this.

For popover specifically, the fault here was in underlying data, not in the calculation. The history for popover is hard to follow, but here's my best attempt to thread together what happened:

1. In May 2024, we started computing popover's status from BCD (https://github.com/web-platform-dx/web-features/pull/1100). At the time, BCD incorrectly reported that Safari fully supported light dismiss ([BCD source at v5.5.28](https://github.com/mdn/browser-compat-data/blob/e0d0d36aedc33194927703d5ddc44b251a666b50/api/HTMLElement.json#L1939-L1941)). We published a status that said popover was Baseline newly available; the calculation was correct, but the upstream data was wrong.

2. In September 2024, BCD was corrected to reflect the incomplete implementation in Safari (https://github.com/mdn/browser-compat-data/pull/24379). When we upgraded the BCD package for web-features, our computation picked that up, [correctly regressing popover to limited availability](https://github.com/web-platform-dx/web-features/pull/1797/files#diff-0376ae895f1ddd65709888ca24e258741972212c1fffaeb5a40f286c411c161b).

   This is the point where this issue would've helped. We couldn't tell consumers of web-features data _why_ we had previously reported popover as Baseline and what was broken about popover in Safari before 18.3. https://github.com/web-platform-dx/web-features/issues/1971 covers this even more specifically, but it's blocked on this more general issue.

4. In January 2025 (last week), BCD was updated to show that light dismiss was fixed in Safari (https://github.com/mdn/browser-compat-data/pull/25787). When we upgraded the BCD package for web-features, our computation picked that up too, [correctly progressing popover to Baseline available](https://github.com/web-platform-dx/web-features/pull/2591/files#diff-0376ae895f1ddd65709888ca24e258741972212c1fffaeb5a40f286c411c161b).

**Comment by bramus** [comment]
Thanks for the clarification @ddbeck; I appreciate the detail that went into it.

**Comment by captainbrosset** [comment]
Here is an example which illustrates the need for partial support and notes.

Last year, I help co-author this post on WebGPU: https://web.dev/blog/webgpu-supported-major-browsers, which celebrated an important milestone: the major browser engines now have some WebGPU support, even if there are still significant gaps.

Looking at caniuse for this feature conveys a similar message: https://caniuse.com/webgpu : shades of green, with lots of notes.

But then, looking at two consumers of the web-features data: webstatus.dev and the explorer tells a very different story:
* https://web-platform-dx.github.io/web-features-explorer/features/webgpu/ Chrome, Firefox, and Edge are all red, no partial support, no notes.
* https://webstatus.dev/features/webgpu?q=webgpu

The data isn't wrong, and the Limited Availability status is correct too. It's just a question of optics. From a developer standpoint, looking at the explorer or webstatus.dev, I'll stay away from WebGPU for sure. However, looking at caniuse, I might get enough information to make me want to start experimenting.

**Comment by mirisuzanne** [comment]
The difference is extremely stark with Anchor Positioning. Caniuse gives it green across the board, while the feature explorer claims there is only Firefox support:

- https://caniuse.com/css-anchor-positioning
- https://web-platform-dx.github.io/web-features-explorer/features/anchor-positioning/

**Comment by jgraham** [comment]
Anchor positioning is a hard enough case that it has its own issue: https://github.com/web-platform-dx/web-features/issues/3558. The claim there is that anyone writing anchor positioning layouts today should expect them to break in the future due to spec changes. If that's true it sounds like a case where web-features is giving authors correct advice and caniuse is not.

Notably it's also different to WebGPU where the differences which web-features can't capture at present are mostly about platform and hardware support. That's something we could handle explicitly without an open-ended notion of "partial" implementation.
